Add instance custom volumes

edlerd opened the pull request:
## Done

- allow attaching custom storage to instances/profiles
- overhaul the instance/profile network configuration

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @lorumic or @edlerd for access.
    - With a local copy of this branch, run as described [here](../HACKING.md#setting-up-for-development).
2. Perform the following QA steps:
    - create/edit instance/profile configuration>storage. Attach storage, Detach storage
    - create/edit instance/profile configuration>network. Attach network, Detach network

webteam-app commented:
Demo starting at https://lxd-ui-463.demos.haus

piperdeck commented:
@edlerd I made a mockup based on your suggestions. There's a screenshot attached, and here's a link to the figma file: https://www.figma.com/file/akl8XNqp2K5egFKI3RIfoG/23.10-LXD---ISOs-and-Storage-Volumes?type=design&node-id=454-1386&mode=design&t=Gh7XPMukUC9MaIrW-0

2 notes:

- I think it might help to change the behaviour of the columns, so that they try to hug the width of their contents a little more. Note how I've shifted the "inherited" and "override" columns way over 
- I've also adjusted the wording of the "attach volume" button to "New storage device", to make it conform with the new name of the config section.

![Frame 608](https://github.com/canonical/lxd-ui/assets/45367207/170d6044-4c77-405e-be09-efa83174d846)

piperdeck reviewed (changes requested):
See previous comment

edlerd commented:
> See previous comment

Updated the layout accordingly and fixed a bunch of bugs. It should be good now. I think with each iteration this is getting better.

Three things to note:
- The inherited disable state should probably make more clear the device is disabled.
- The volume creation on adding a new device to an instance (in the modal) is a bit wired still
- The spacing in the instance devices sections seems partly off.

piperdeck commented:
OK, so I sought some advice from Ana about this, and looked into how table spacing works in MAAS, and I came up with another iteration. I think this addresses a lot of the challenges previously mentioned, and makes the layout a lot easier to read.

Note that my use of padding was very intentional in this prototype, so I would recommend using Dev Mode to inspect the layout of the containers to see how much white space I've used to convey hierarchy.

Ping me with any questions you have, and I can answer them on Tuesday.

Link to figma: https://www.figma.com/file/akl8XNqp2K5egFKI3RIfoG/23.10-LXD---ISOs-and-Storage-Volumes?type=design&node-id=472-238&mode=design

piperdeck reviewed (changes requested):
See comment

edlerd commented:
> OK, so I sought some advice from Ana about this, and looked into how table spacing works in MAAS, and I came up with another iteration. I think this addresses a lot of the challenges previously mentioned, and makes the layout a lot easier to read.

Yes, this is much cleaner than before. All changes should be implemented. Please give it another round of review @piperdeck .

piperdeck reviewed (changes requested):
This is looking almost perfect

# 1

It seems like when you change the root pool of an instance, it will error out. I'm sure there's some situation where it's possible, but I think the least we can do to avoid this error would be to have the "pool" dropdown default to the inherited value when you first create an override

![image](https://github.com/canonical/lxd-ui/assets/45367207/718c658f-106b-40ad-a160-bf0d3722ffe3)

# 2

For the custom devices section, the baselines of the keys and values are nicely aligned until you go into edit mode. I think this would look a lot nicer if the baselines could remain aligned

Read mode:
![image](https://github.com/canonical/lxd-ui/assets/45367207/512b28d2-1d6b-4f21-8f07-1cfd63107eca)

Edit mode:
![image](https://github.com/canonical/lxd-ui/assets/45367207/144eaec7-4c95-4012-b84e-1dec0d020604)

# 3

After comparing with the network devices tab, and seeing as the name of the tab is "disk devices", I think it might be better if the button to add a custom volume was changed to "attach disk device"

edlerd commented:
Thanks for the review @piperdeck -- I added your good suggestions, please have another look if everything is now as expected.

piperdeck commented:
Thank you for adjusting the spacing of those elements, they look a lot better in edit mode now.

However, I'm not super keen on center-aligning the inherited configs. I think it would look better if the rows always had their text baselines aligned, and using the topmost row if a given cell has two rows of text. Examples:

![image](https://github.com/canonical/lxd-ui/assets/45367207/2e4343d5-e4f8-4e88-a9c9-b17adab33b9d)

![image](https://github.com/canonical/lxd-ui/assets/45367207/ff16e8f3-b848-4feb-936f-85984b39f37a)

![image](https://github.com/canonical/lxd-ui/assets/45367207/6aa6c2c8-6fa6-41a7-a6f5-964bb7aa9831)

![image](https://github.com/canonical/lxd-ui/assets/45367207/813dac7d-4f68-4918-a148-7d7c237d64cd)

piperdeck reviewed (changes requested):
See comments

edlerd commented:
> However, I'm not super keen on center-aligning the inherited configs. I think it would look better if the rows always had their text baselines aligned, and using the topmost row if a given cell has two rows of text. Examples:

Took some care and custom CSS to make the device page rows aligned in edit mode and read mode.

piperdeck reviewed (approved):
Looks great!

lorumic commented on src/pages/instances/forms/DiskDeviceFormCustom.tsx:
```diff
@@ -0,0 +1,260 @@
+              <b>{formVolume.path}</b>
+            </div>
+          ) : (
+            <Input
+              id={`pool-${index}-path`}
+              onBlur={formik.handleBlur}
+              onChange={(e) => {
+                formik.setFieldValue(`devices.${index}.path`, e.target.value);
+              }}
+              value={formVolume.path}
+              type="text"
+              placeholder="Enter full path (e.g. /data)"
+              className="u-no-margin--bottom"
+            />
```
I think this field should be `required`, to avoid the postponed feedback in case the user forgets to write the path and tries to save the changes:

![image](https://github.com/canonical/lxd-ui/assets/56583786/4226497f-5155-4d13-a261-7700d67c1233)

edlerd replied:
Made the two labels required. We should add some logic to prevent submitting the forms with missing values in the future. But that is a bit more involved, so lets keep it as part of this issue https://github.com/canonical/lxd-ui/issues/485

lorumic replied:
Yeah just marking the field as required doesn't do much indeed, we need to be able to properly validate these required fields so that they prevent the user from clicking the submit button if they contain invalid values. Let's remember this as a future improvement - out of scope for now.

lorumic reviewed (commented):
Didn't do a full review yet, found the below issue on first pass.

lorumic commented on src/pages/instances/forms/NetworkDevicesForm.tsx:
```diff
@@ -135,9 +136,22 @@ or remove the originating item"
+                  {isReadOnly || device.type === "custom-nic" ? (
+                    device.name
+                  ) : (
+                    <Input
+                      label="Device name"
+                      name={`devices.${index}.name`}
+                      id={`networkName${index}`}
+                      onBlur={formik.handleBlur}
+                      onChange={formik.handleChange}
+                      value={
+                        (formik.values.devices[index] as LxdNicDevice).name
+                      }
+                      type="text"
+                      placeholder="Enter name"
+                    />
```
Same as above:

![image](https://github.com/canonical/lxd-ui/assets/56583786/b1c8817c-c143-4b1c-8d69-5c4a2acd8b8c)

lorumic commented on src/pages/instances/forms/DiskDeviceForm.tsx:
I wonder if this could be a good chance to move this (and the other shared components) outside of the `instances` directory tree, since it is kind of shared between instances and profiles, and it seems arbitrary to force it inside this directory tree instead of the other.

Anyway, I'm really happy that this is finally reaching the form that I was envisioning around 11 months ago, when we discussed the storage devices for the first time in the profile creation form. Back then we decided to start with root storage only, disallowing any other disk device configuration (except with the YAML form), and now we finally have this configurable in the form as well!
This is a nice step forward that I want to celebrate. It was a long run, but it's getting closer and closer to feature parity with the CLI. :partying_face:

lorumic replied:
My better idea was the option one that you said you don't want to do. I prefer that as the current status - it's probably clearer and less confusing. There is no reason to have these components under the instances tree, it's just arbitrary. Why not under profiles then?

edlerd replied:
We could have a shared module for this, actually there are some components like the `YamlForm`, which are used in even more modules. And it already happens, that the `CloudInitConfig` in the profiles module, though used from instances, while all other reused form components are under the instances module.
I'll move them into a new `forms` folder under the shared `components`.

lorumic commented on src/types/device.d.ts:
```diff
@@ -1,8 +1,11 @@
   pool: string;
   size?: string;
+  source?: string;
+  "limits.read"?: string;
+  "limits.write"?: string;
```
Maybe we can have:
```suggestion
  limits?: {
    read?: string;
    write?: string;
  };
```
?

edlerd replied:
No, we can't change it, because this is what LXD expects in the payload and how it is being encoded in JSON.

lorumic commented on src/util/formDevices.tsx:
```diff
@@ -29,10 +35,19 @@ export interface CustomNetworkDevice {
 
+export type FormDiskDevice = Partial<LxdDiskDevice> &
+  Required<Pick<LxdDiskDevice, "name" | "pool">> & {
+    limits?: {
+      read?: string;
+      write?: string;
+    };
```
With the change proposed above, this should have now become part of the `LxdDiskDevice` type which is already expressed by `Partial<LxdDiskDevice>`, so I think it could be removed from here.

edlerd replied:
We have to keep the structure for LXD different, but as formik doesn't allow keys to contain a `.`, we use this different structure for the form type.

lorumic commented on src/util/instanceConfigInheritance.tsx:
```diff
@@ -131,12 +131,47 @@ export const figureInheritedRootStorage = (
+        continue;
       }
+      Object.entries(profile.devices)
+        .filter(([key, network]) => isDiskDevice(network) && key !== "root")
```
Huh? This looks a bit confusing. Maybe:
```suggestion
        .filter(([key, device]) => isDiskDevice(device) && key !== "root")
```

lorumic reviewed (commented):
This looks and behaves as expected, I added some comments for a few minor improvements.

edlerd commented on src/components/forms/RenameDiskDeviceInput.tsx:
```diff
@@ -0,0 +1,58 @@
+  setName,
+  isReadOnly,
+}) => {
+  const [iseEditing, setIsEditing] = React.useState(false);
```
Yes, though we need to also take care of the references. Will fix it locally.

edlerd commented on src/types/device.d.ts:
```diff
@@ -1,8 +1,11 @@
 export interface LxdDiskDevice {
   name?: string;
-  path: string;
+  path?: string;
```
It is indeed optional, when the Volume is of type `block` we must not specify a path. Only with volume type `filesystem` a path is mandatory. Unfortunately, we have no access to the chosen volume type here, hence we need to be ambiguous about whether path is mandatory or not.

edlerd commented on src/pages/storage/CustomVolumeSelectModal.tsx:
```diff
@@ -0,0 +1,141 @@
+        </ScrollableTable>
+      </div>
+      <footer className="p-modal__footer" id="modal-footer">
+        <Button className="u-no-margin--bottom" onClick={onCancel}>
```
I think it makes sense here, in other cases the 2nd button is positive or negative, so a default button is fine in my opinion. Will take a note to discuss this with design in the future.

lorumic replied:
This too can be a positive button in some cases. Actually, I just noticed that by default it gets the positive appearance at first load - namely, when we're still fetching the volumes. Is this expected and desired? Or should it have the default appearance by default and only take the positive one when we've fetched the list and there's no custom volume available?

edlerd replied:
True, there is a little glitch here while loading. I think we shouldn't render the buttons while loading, will change it.

edlerd replied:
Give it another default appearance before loading wouldn't solve it, as depending on the context, it can have both values after loading.

lorumic replied:
Seeing it in action, though, makes me wonder if this is acceptable from a design perspective, because it's changing the modal's height - which is something that was mentioned many times before, that we should not do. Maybe visible but disabled buttons would be a better choice? Might be worth investigating this with @piperdeck

edlerd replied:
Creation should be spaced now as expected.

lorumic replied:
Not sure if we care, but it still looks bad at >1020px height  - this for example is with a ~1280px height:

![image](https://github.com/canonical/lxd-ui/assets/56583786/c937c8bc-2a90-48e3-aa7e-9d82dffeb90a)

edlerd replied:
That's not what should be happening. I think we are getting closer. Now it should work there as well.

lorumic replied:
Yes, I confirm that now it works well at any resolution. :+1:

edlerd commented on src/pages/instances/forms/DiskDeviceFormCustom.tsx:
```diff
@@ -0,0 +1,262 @@
+
+    if (formVolume.path !== undefined) {
+      rows.push(
+        getConfigurationRowBase({
```
I took note of this to improve on it in the future.
